swayidle/main.c

```diff
@@ -432,10 +433,13 @@ int main(int argc, char *argv[]) {
 	struct wl_event_source *source = wl_event_loop_add_fd(state.event_loop,
 		wl_display_get_fd(state.display), WL_EVENT_READABLE,
 		display_event, NULL);
-	wl_event_source_check(source);
```

Comment:
I don't think this is right.

Reply:
I pushed a different fix, namely changing `wl_display_dispatch` to `wl_display_dispatch_pending` so that it doesn't block.

Reply:
I think removing this line is right now, because it's causing double `display_event` each time.

Reply:
This is needed if an event happens somewhere else (dbus, signal) and sends Wayland messages.

Reply:
But then why won't they cause an event on polling first time round?

The alternative is to check whether `read_wl_events` has been set in `display_event` to prevent trying to read again.

Reply:
Actually I think we're not using these functions correctly. I just read how Weston does it for its Wayland backend. The event loop is just `wl_event_loop_dispatch` in a loop. The `display_event` handler uses the mask to know what to do: https://gitlab.freedesktop.org/wayland/weston/blob/master/libweston/compositor-wayland.c#L2593

They also use `wl_event_source_check`.

Can you try doing the same?

>But then why won't they cause an event on polling first time round?

Because "I need to send data on the Wayland socket" is a consequence of "I've handled an D-Bus event", and the event loop isn't aware of that without `wl_event_source_check`.

Reply:
I did it like in the link, though I omitted checking for `WL_EVENT_WRITABLE` because I thought that we're only checking for readable events. Is my understanding incorrect?
Also, I kept the count variable in, that's correct, right?

Reply:
Yes, because https://github.com/swaywm/sway/blob/eda3bfeed5097c71634332bfe998188b028abf02/swayidle/main.c#L433 means it only triggers on readable events? or is my thinking wrong?

Reply:
Ah, right. We should add `| WL_EVENT_WRITABLE` then. It probably doesn't happen right now because we don't fill the receiver buffers, but it's probably best to handle it properly.

Reply:
https://gitlab.freedesktop.org/wayland/weston/issues/163